modules/packages/EpochManager.chpl

```diff
@@ -0,0 +1,1147 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+
+To use the :class:`EpochManager`, first create an instance.
+
+.. code-block:: chapel
+
+ var manager = new unmanaged EpochManager();
+
+
+Registering a Task
+------------------
+A task must be registered with the manager in order to use the manager.
+Registration returns a token.
+
+.. code-block:: chapel
+
+ var tok = manager.register();
+
+
+Pinning/Unpinning a Task
+------------------------
+To avoid reclamation while a task is accessing a resource, I.E. to enter
+critical section, a task must `pin`. Correspondingly to exit critical section,
+the task must `unpin`.
+
+.. code-block:: chapel
+
+ // Enter critical section
+ tok.pin();
+ // Do something
+
+ // Exit critical section
+ tok.unpin();
+
+
+Deleting an object
+------------------
+To `delete` an object:
+
+.. code-block:: chapel
+
+ tok.delete_obj(myObj);
+
+
+.. note::
+ A task must be `pinned` to `delete` an object. The manager can only be used to
+ delete ``unmanaged`` objects.
```

Comment:
If this is the case, why doesn't `tok.pin()` return an object that then you can call `deleteObj` on?

Reply:
My question would be 'why'? Right now, you need to obtain a `token` via registration; having to get another object to invoke `deleteObj` on rather than just using the token just seems like its adding extra work without any benefit; this object would also need to be heap-allocated.

Reply:
The benefit would be that the implementation would enforce the constraint noted above.